File: demo/demo_LATROT.py
# N. M. Rathmann <rathmann@nbi.ku.dk> and D. A. Lilien <dlilien90@gmail.com>, 2020

import sys, os, copy, code # code.interact(local=locals())
import numpy as np
import scipy.special as sp
from netCDF4 import Dataset
from specfabpy import specfabpy as sf 

#### NOTICE ####
# To use "specfabpy", you must first compile the Python module by running "make specfabpy" (compiles the F2PY wrapper for "specfab.f90")
################

#----------------------
# Input arguments
#----------------------

if len(sys.argv) != 2:
    print('usage: python3 %s ugrad'%(sys.argv[0]))
    sys.exit(0)
    
# Arguments
arg_exp = sys.argv[1]    # Velocity gradient experiment

#----------------------
# Numerics
#----------------------

Nt = 230  # Number of integration points
dt = 0.02 # Time-step size

L = 12 # Spectral truncation
nu0 = 5.0e-3 # Regularization strength (calibrated for L)

#----------------------
# Grain parameters
#----------------------

# Optimal n'=1 (lin) grain parameters 
# These are the linear mixed Taylor--Sachs best-fit parameters from Rathmann and Lilien (2021)
Eca_opt_lin   = 1e3
Ecc_opt_lin   = 1e0
alpha_opt_lin = 0.0125

# Optimal n'=3 (nlin) grain parameters 
# These are the nonlinear Sachs-only best-fit parameters (Rathmann et. al, 2021) 
Eca_opt_nlin   = 1e2
Ecc_opt_nlin   = 1e0
alpha_opt_nlin = 0

#----------------------
# Velocity gradient
#----------------------

# "ugrad" is the specified (constant) velocity gradient.

ugrad = np.diag([0.5, 0.5, -1.0]) 

if arg_exp == 'uc_xx': ugrad = np.diag([-1, .5, .5])
if arg_exp == 'uc_yy': ugrad = np.diag([.5, -1, .5])
if arg_exp == 'uc_zz': ugrad = np.diag([.5, .5, -1])

if arg_exp == 'ue_xx': ugrad = -1*np.diag([-1, .5, .5])
if arg_exp == 'ue_yy': ugrad = -1*np.diag([.5, -1, .5])
if arg_exp == 'ue_zz': ugrad = -1*np.diag([.5, .5, -1])

if arg_exp == 'cc_zx': ugrad = np.diag([1, 0, -1])
if arg_exp == 'cc_zy': ugrad = np.diag([0, 1, -1])
if arg_exp == 'cc_yx': ugrad = np.diag([1, -1, 0])

if arg_exp == 'ss_xz': ugrad = np.array([[0,0,1], [0,0,0], [0,0,0]])
if arg_exp == 'ss_xy': ugrad = np.array([[0,1,0], [0,0,0], [0,0,0]])
if arg_exp == 'ss_yz': ugrad = np.array([[0,0,0], [0,0,1], [0,0,0]])

if arg_exp == 'rr_xz': ugrad = np.array([[0,0,+1], [0,0,0], [-1,0,0]])
if arg_exp == 'rr_xy': ugrad = np.array([[0,+1,0], [-1,0,0], [0,0,0]])
if arg_exp == 'rr_yz': ugrad = np.array([[0,0,0], [0,0,+1], [0,-1,0]])

eps = (ugrad+np.transpose(ugrad))/2 # Symmetric part (strain-rate)
omg = (ugrad-np.transpose(ugrad))/2 # Anti-symmetric part (spin)

#----------------------
# Initialize model
#----------------------

nlm_len = sf.init(L) # nlm_len is the number of fabric expansion coefficients (degrees of freedom).
nlm = np.zeros((Nt,nlm_len), dtype=np.complex128) # The expansion coefficients
lm = sf.get_lm(nlm_len) # The (l,m) values corresponding to the coefficients in "nlm".

# Initial fabric state
if arg_exp[0:2] == 'rr':
    # Init with some anisotropy
    nlm[0,0] = np.sqrt(1/2 + 0j)
    nlm[0,2] = np.sqrt(1/2 + 0j)
else:
    # Init with isotropy
    nlm[0,0] = 1/np.sqrt(4*np.pi) # Normalized such that N(t=0) = 1

#----------------------
# Integrate
#----------------------

print('Numerics: Nt=%i, dt=%f, L=%i (nlm_len=%i)'%(Nt,dt,L,nlm_len))

tau,Aprime,Ecc,Eca,beta = 0*eps,0,1,1,1 # Fabric evolves by a plastic spin that depends on large-scale velocity gradient only (Taylor style => beta = 1).
dndt_ij_ROT = sf.get_dndt_ij_latrot(nlm_len, eps,omg, tau,Aprime,Ecc,Eca,beta) # Here we consider constant large-scale velocity gradients, **but if for any practical time-varying scenario dndt_ij should be calculated inside the loop below!**
dndt_ij_REG = sf.get_dndt_ij_reg(nlm_len)
nu = sf.get_nu_eps(nu0, eps)
dndt_ij = dndt_ij_ROT + nu*dndt_ij_REG
 
# Euler integration scheme
for tt in np.arange(1,Nt):
    nlm[tt,:] = nlm[tt-1,:] + dt*np.tensordot(dndt_ij, nlm[tt-1,:], axes=(-1,0))

#----------------------
# Aux state vars
#----------------------

#### Relevant only for calculating the enhancement-factors resulting from the fabric.

###if arg_nprime == 1: nprime, Ecc, Eca = 1, 1, 1e4
###if arg_nprime == 3: nprime, Ecc, Eca = 3, 1, 1e2
###alpha = 0 # Sachs--Taylor weight: 0 = 100% Sachs, 1 = 100% Taylor

# Calculate eigenvalues, principal directions, and enhancement-factors

vecdim = (Nt,3)
eigvals = np.zeros(vecdim, dtype=np.float64)
Eeiej_lin,Eeiej_nlin = np.zeros((Nt,3,3), dtype=np.float64), np.zeros((Nt,3,3), dtype=np.float64)
Epijqij_lin, Epijqij_nlin = np.zeros(vecdim, dtype=np.float64), np.zeros(vecdim, dtype=np.float64)
e1,e2,e3    = np.zeros(vecdim, dtype=np.float64),np.zeros(vecdim, dtype=np.float64),np.zeros(vecdim, dtype=np.float64)
p23,p12,p13 = np.zeros(vecdim, dtype=np.float64),np.zeros(vecdim, dtype=np.float64),np.zeros(vecdim, dtype=np.float64)
q23,q12,q13 = np.zeros(vecdim, dtype=np.float64),np.zeros(vecdim, dtype=np.float64),np.zeros(vecdim, dtype=np.float64)

for tt in np.arange(0,Nt):
    c = nlm[tt,:]
    
    e1[tt,:],e2[tt,:],e3[tt,:], eigvals[tt,:] = sf.get_eigenframe(c)
    p23[tt,:],p12[tt,:],p13[tt,:], q23[tt,:],q12[tt,:],q13[tt,:] = sf.get_pqframe(c)

    # Linear (n'=1) mixed Taylor--Sachs enhancements            
    nprime = 1
    Eeiej_lin[tt,:,:] = np.transpose(sf.get_eeiej(c, Ecc_opt_lin, Eca_opt_lin, alpha_opt_lin, nprime))
    Epijqij_lin[tt,:] = sf.get_epijqij(c, Ecc_opt_lin, Eca_opt_lin, alpha_opt_lin, nprime)
    
    # Nonlinear (n'=3) Sachs enhancements
    nprime = 3
    Eeiej_nlin[tt,:,:] = np.transpose(sf.get_eeiej(c, Ecc_opt_lin, Eca_opt_lin, alpha_opt_lin, nprime))
    Epijqij_nlin[tt,:] = sf.get_epijqij(c, Ecc_opt_lin, Eca_opt_lin, alpha_opt_lin, nprime)
    
#----------------------
# Save
#----------------------

# Save the solution to a netCDF file.
# You can plot the results stored in the netCDF file using "plot.py".

fname = 'solutions/LATROT_%s.nc'%(arg_exp)
ncfile = Dataset(fname,mode='w',format='NETCDF3_CLASSIC') 

# Config
ncfile.tsteps, ncfile.dt, ncfile.nu, ncfile.L = Nt, dt, nu, L
ncfile.Ecc_opt_lin,  ncfile.Eca_opt_lin,  ncfile.alpha_opt_lin  = Ecc_opt_lin,  Eca_opt_lin,  alpha_opt_lin
ncfile.Ecc_opt_nlin, ncfile.Eca_opt_nlin, ncfile.alpha_opt_nlin = Ecc_opt_nlin, Eca_opt_nlin, alpha_opt_nlin

ncfile.ugrad = ugrad.flatten()

# Dimensions
c_did       = ncfile.createDimension('DOF',     nlm_len)
time_did    = ncfile.createDimension('tstep',   Nt)
eig_did     = ncfile.createDimension('eigval',  3)
dim_did     = ncfile.createDimension('dim',     3)
pair_did    = ncfile.createDimension('pair',    2)

# Variables
myint = np.int32
myflt = np.float64
dimarr_vec = ('lm','lm','lmdyn')

f_lm   = ncfile.createVariable('lm', myint, ('DOF','pair')) 
f_c_re = ncfile.createVariable('c_re', myflt, ('tstep','DOF')) 
f_c_im = ncfile.createVariable('c_im', myflt, ('tstep','DOF')) 
mkvec = lambda field: ncfile.createVariable(field, myflt, ('tstep','dim'))
f_eigvals = ncfile.createVariable('eigvals', myflt, ('tstep','eigval'))

f_Eeiej_lin    = ncfile.createVariable('Eeiej_lin',   myflt, ('tstep','dim','dim'))
f_Epijqij_lin  = ncfile.createVariable('Epijqij_lin', myflt, ('tstep','dim'))
f_Eeiej_nlin   = ncfile.createVariable('Eeiej_nlin',   myflt, ('tstep','dim','dim'))
f_Epijqij_nlin = ncfile.createVariable('Epijqij_nlin', myflt, ('tstep','dim'))

f_e1, f_e2, f_e3    = mkvec('e1'),  mkvec('e2'),  mkvec('e3')
f_p23, f_p12, f_p13 = mkvec('p23'), mkvec('p12'), mkvec('p13')
f_q23, f_q12, f_q13 = mkvec('q23'), mkvec('q12'), mkvec('q13')

f_lm[:,:] = lm.T
f_c_re[:,:], f_c_im[:,:] = np.real(nlm), np.imag(nlm)
f_eigvals[:,:] = eigvals
f_Eeiej_lin[:,:,:] = Eeiej_lin
f_Epijqij_lin[:,:] = Epijqij_lin
f_Eeiej_nlin[:,:,:] = Eeiej_nlin
f_Epijqij_nlin[:,:] = Epijqij_nlin
f_e1[:,:], f_e2[:,:], f_e3[:,:]    = e1, e2, e3
f_p23[:,:], f_p12[:,:], f_p13[:,:] = p23, p12, p13
f_q23[:,:], f_q12[:,:], f_q13[:,:] = q23, q12, q13

ncfile.close(); 
print('Solution dumped in %s'%fname)
print('Plot result:\npython3 plot_demo_LATROT.py %s'%(arg_exp))

File: demo/demo_LATROT.f90
! N. M. Rathmann <rathmann@nbi.ku.dk> and D. A. Lilien <dlilien90@gmail.com>, 2020

program demo

    use specfab    
    use netcdf
    
    implicit none

    integer, parameter :: dp = 8

    ! Numerics
    real, parameter    :: dt = 0.02 ! Time-step size
    integer, parameter :: Nt = 230  ! Number of time steps
    integer            :: Lcap = 12     ! Expansion series truncation
    real(kind=dp)      :: nu0  = 5.0d-3 ! Regularization magnitude calibrated for demo with L=12
    
    ! Constants and argv strings    
    integer :: ii,tt ! loop vars
    character(len=5) :: arg_exp ! experiment type (see below)

    ! Fabric state and evolution
    complex(kind=dp), allocatable :: nlm(:), dndt(:,:), dndt_ROT(:,:), dndt_REG(:,:) ! Series expansion coefs and evolution matrix
    real(kind=dp) :: ugrad(3,3), eps(3,3), omg(3,3) ! Large-scale deformation

    ! For dumping state to netCDF
    complex(kind=dp), allocatable   :: nlm_save(:,:)
    real(kind=dp)                   :: eigvals_save(3,Nt), Eeiej_lin_save(3,3,Nt), Eeiej_nlin_save(3,3,Nt), Epijqij_lin_save(3,Nt), Epijqij_nlin_save(3,Nt)
    real(kind=dp), dimension(3,Nt)  :: e1_save,e2_save,e3_save, p23_save,p12_save,p13_save, q23_save,q12_save,q13_save
    character(len=30) :: fname_sol
    integer :: ncid, c_did, time_did, eig_did, dim_did, pair_did ! Dimension IDs
    integer :: id_cre,id_cim,id_lm, id_eig, id_e1,id_e2,id_e3, id_p23,id_p12,id_p13, id_q23,id_q12,id_q13 ! Var IDs
    integer :: id_Eeiej_lin, id_Eeiej_nlin, id_Epijqij_lin, id_Epijqij_nlin ! Var IDs

    if (command_argument_count() .ne. 1) then
        print *,'usage: ./demo ugrad'
        call exit(0)
    end if

    !-------------------------------------------------------------------
    ! Velocity gradient tensor
    !-------------------------------------------------------------------

    call get_command_argument(1, arg_exp)
    select case (arg_exp)

        ! RECALL COLUMN FIRST IN FORTRAN
        
        ! Uniaxial compression (uc) and uniaxial extension (ue)
        case ('uc_xx', 'ue_xx')
            ugrad = reshape([-1.0,0.,0., 0.,0.5,0., 0.,0.,0.5], [3, 3])
        case ('uc_yy', 'ue_yy')
            ugrad = reshape([0.5,0.,0., 0.,-1.0,0., 0.,0.,0.5], [3, 3])
        case ('uc_zz', 'ue_zz')
            ugrad = reshape([0.5,0.,0., 0.,0.5,0., 0.,0.,-1.], [3, 3])

        ! Confined compression (cc)
        case ('cc_zx')
            ugrad = reshape([1.,0.,0., 0.,0.,0., 0.,0.,-1.], [3, 3]) ! confined in y
        case ('cc_zy')
            ugrad = reshape([0.,0.,0., 0.,1.,0., 0.,0.,-1.], [3, 3]) ! confined in x
        case ('cc_yx')
            ugrad = reshape([1.,0.,0., 0.,-1.,0., 0.,0.,0.], [3, 3]) ! confined in z
                    
        ! Simple shear (ss)
        case ('ss_xz')
            ugrad = reshape([0.,0.,0., 0.,0.,0., 1.,0.,0.], [3, 3]) 
        case ('ss_xy')
            ugrad = reshape([0.,0.,0., 1.,0.,0., 0.,0.,0.], [3, 3]) 
        case ('ss_yz')
            ugrad = reshape([0.,0.,0., 0.,0.,0., 0.,1.,0.], [3, 3]) 

        ! Ridig rotation (rr)
        case ('rr_xz')
            ugrad = reshape([0.,0.,-1., 0.,0.,0., +1.,0.,0.], [3, 3])
        case ('rr_xy')
            ugrad = reshape([0.,-1.,0., +1.,0.,0., 0.,0.,0.], [3, 3]) 
        case ('rr_yz')
            ugrad = reshape([0.,0.,0., 0.,0.,-1., 0.,+1.,0.], [3, 3]) 

        case default
            print *,'argv error: valid "ugrad" are "uc_ij", "ue_ij", "cc_ij", "ss_ij", "rr_ij" where i,j are any of x,y,z'
            call exit(0)
    end select

    select case (arg_exp)
        case ('ue_xx','ue_yy','ue_zz')
            ugrad = -1*ugrad
    end select

    eps = (ugrad+transpose(ugrad))/2 ! strain-rate
    omg = (ugrad-transpose(ugrad))/2 ! spin
            
    !-------------------------------------------------------------------
    ! Initialize
    !-------------------------------------------------------------------

    call initspecfab(Lcap) ! nlm_len is now defined (number of expansion coeffcients, i.e. #DOFs)
    allocate(nlm(nlm_len))
    nlm = [(0,ii=1,nlm_len)] ! Expansion coefs "n_l^m" are saved in the 1D array "nlm". Corresponding (l,m) values for the i'th coef (i.e. nlm(i)) are (l,m) = (lm(1,i),lm(2,i))
    allocate(nlm_save(nlm_len,Nt))
    allocate(dndt(nlm_len,nlm_len))
    allocate(dndt_ROT(nlm_len,nlm_len))
    allocate(dndt_REG(nlm_len,nlm_len))    

    
    select case (arg_exp)
        case ('rr_xz','rr_xy','rr_yz')
            ! Init with some anisotropy
            nlm(1) = sqrt(cmplx(1/2.,0))
            nlm(3) = sqrt(cmplx(1/2.,0))
        case default
            ! Init with isotropy
            nlm(1) = (1,0)
            nlm(1) = nlm(1)/f_ev_c0(nlm(1)) ! Normalize
    end select
    
    write(*,"(A13,I4,A5,F12.10,A4,I2,A10,I3,A1)") 'Numerics: Nt=', Nt, ', dt=', dt, ', L=', Lcap, ' (nlm_len=',nlm_len,')'

    !-------------------------------------------------------------------
    ! Integrate
    !-------------------------------------------------------------------

    call savestate(nlm, 1) ! Save initial state    
    dndt_ROT = dndt_ij_LATROT(eps,omg, 0*eps,0d0,0d0,0d0, 1d0) ! Assume constant strain-rate and spin with Taylor style plastic spin for lattice rotation (beta=1).
    dndt_REG = f_nu_eps(nu0, eps) * dndt_ij_REG() ! Regularization: nu * (reg. mag.) *  reg. matrix
            
    do tt = 2, Nt
!        write(*,"(A9,I3)") '*** Step ', tt
        dndt = dndt_ROT + dndt_REG 
        nlm = nlm + dt * matmul(dndt, nlm) ! Spectral coefficients evolve by a linear transformation
        call savestate(nlm, tt)
    end do
    
    !-------------------------------------------------------------------
    ! Dump solution to netCDF
    !-------------------------------------------------------------------
    
    write (fname_sol,"('solutions/ROT_',A5,'.nc')") arg_exp
    call check( nf90_create(fname_sol, NF90_CLOBBER, ncid) )
    
    call check(nf90_put_att(ncid,NF90_GLOBAL, "tsteps", Nt))
    call check(nf90_put_att(ncid,NF90_GLOBAL, "dt",     dt))
    call check(nf90_put_att(ncid,NF90_GLOBAL, "nu",     f_nu_eps(nu0, eps) ))
    call check(nf90_put_att(ncid,NF90_GLOBAL, "L",      Lcap))
    call check(nf90_put_att(ncid,NF90_GLOBAL, "ugrad",  reshape(ugrad, [size(ugrad)]) ))
    
    call check(nf90_put_att(ncid,NF90_GLOBAL, "Eca_opt_lin", Eca_opt_lin))
    call check(nf90_put_att(ncid,NF90_GLOBAL, "Ecc_opt_lin", Ecc_opt_lin))
    call check(nf90_put_att(ncid,NF90_GLOBAL, "Eca_opt_nlin", Eca_opt_nlin))
    call check(nf90_put_att(ncid,NF90_GLOBAL, "Ecc_opt_nlin", Ecc_opt_nlin))
    call check(nf90_put_att(ncid,NF90_GLOBAL, "alpha_opt_lin",  alpha_opt_lin))
    call check(nf90_put_att(ncid,NF90_GLOBAL, "alpha_opt_nlin", alpha_opt_nlin))
    
    call check( nf90_def_dim(ncid, "DOF",    nlm_len,   c_did) )
    call check( nf90_def_dim(ncid, "tstep",  Nt,        time_did) )
    call check( nf90_def_dim(ncid, "eigval", 3,         eig_did) )
    call check( nf90_def_dim(ncid, "dim",    3,         dim_did) )
    call check( nf90_def_dim(ncid, "pair",   2,         pair_did) )

    call check( nf90_def_var(ncid, "lm",    NF90_INT,    [pair_did, c_did], id_lm) )
    call check( nf90_def_var(ncid, "c_re",  NF90_DOUBLE, [c_did,   time_did], id_cre) )
    call check( nf90_def_var(ncid, "c_im",  NF90_DOUBLE, [c_did,   time_did], id_cim) )
    
    call check( nf90_def_var(ncid, "eigvals", NF90_DOUBLE, [eig_did, time_did], id_eig) )
    call check( nf90_def_var(ncid, "Eeiej_nlin",   NF90_DOUBLE, [dim_did,dim_did, time_did], id_Eeiej_nlin) )
    call check( nf90_def_var(ncid, "Eeiej_lin",    NF90_DOUBLE, [dim_did,dim_did, time_did], id_Eeiej_lin) )
    call check( nf90_def_var(ncid, "e1",      NF90_DOUBLE, [dim_did, time_did], id_e1) )
    call check( nf90_def_var(ncid, "e2",      NF90_DOUBLE, [dim_did, time_did], id_e2) )
    call check( nf90_def_var(ncid, "e3",      NF90_DOUBLE, [dim_did, time_did], id_e3) )
    call check( nf90_def_var(ncid, "Epijqij_nlin", NF90_DOUBLE, [dim_did, time_did], id_Epijqij_nlin) )
    call check( nf90_def_var(ncid, "Epijqij_lin",  NF90_DOUBLE, [dim_did, time_did], id_Epijqij_lin) )
    call check( nf90_def_var(ncid, "p23",     NF90_DOUBLE, [dim_did, time_did], id_p23) )
    call check( nf90_def_var(ncid, "p12",     NF90_DOUBLE, [dim_did, time_did], id_p12) )
    call check( nf90_def_var(ncid, "p13",     NF90_DOUBLE, [dim_did, time_did], id_p13) )
    call check( nf90_def_var(ncid, "q23",     NF90_DOUBLE, [dim_did, time_did], id_q23) )
    call check( nf90_def_var(ncid, "q12",     NF90_DOUBLE, [dim_did, time_did], id_q12) )
    call check( nf90_def_var(ncid, "q13",     NF90_DOUBLE, [dim_did, time_did], id_q13) )
    
    call check( nf90_enddef(ncid) )
    
    call check( nf90_put_var(ncid, id_cre,   real(nlm_save)) )
    call check( nf90_put_var(ncid, id_cim,   aimag(nlm_save)) )
    call check( nf90_put_var(ncid, id_lm,    lm(:,1:nlm_len)) )
    
    call check( nf90_put_var(ncid, id_eig,  eigvals_save) )
    call check( nf90_put_var(ncid, id_Eeiej_lin,  Eeiej_lin_save) )
    call check( nf90_put_var(ncid, id_Eeiej_nlin, Eeiej_nlin_save) )
    call check( nf90_put_var(ncid, id_e1,   e1_save) )
    call check( nf90_put_var(ncid, id_e2,   e2_save) )
    call check( nf90_put_var(ncid, id_e3,   e3_save) )
    call check( nf90_put_var(ncid, id_Epijqij_lin,  Epijqij_lin_save) )
    call check( nf90_put_var(ncid, id_Epijqij_nlin, Epijqij_nlin_save) )
    call check( nf90_put_var(ncid, id_p23,  p23_save) )
    call check( nf90_put_var(ncid, id_p12,  p12_save) )
    call check( nf90_put_var(ncid, id_p13,  p13_save) )
    call check( nf90_put_var(ncid, id_q23,  q23_save) )
    call check( nf90_put_var(ncid, id_q12,  q12_save) )
    call check( nf90_put_var(ncid, id_q13,  q13_save) )
    
    call check( nf90_close(ncid) )

    print *, 'Solution dumped in ', fname_sol
    print *, "Plot result:"
    write(*,"(A25,A5)") "python3 plot_demo_LATROT.py ", arg_exp

contains

    subroutine savestate(nlm, tt)
    
        implicit none
    
        complex(kind=dp), intent(in) :: nlm(nlm_len)
        integer, intent(in) :: tt
    
        nlm_save(:,tt)  = nlm

        call eigenframe(nlm, e1_save(:,tt),e2_save(:,tt),e3_save(:,tt), eigvals_save(:,tt)) 
        call pqframe(nlm, p23_save(:,tt),p12_save(:,tt),p13_save(:,tt), q23_save(:,tt),q12_save(:,tt),q13_save(:,tt))

        ! Linear (n'=1) mixed Taylor--Sachs enhancements        
        Eeiej_lin_save(:,:,tt) = Eeiej(  nlm, Ecc_opt_lin, Eca_opt_lin, alpha_opt_lin, 1)
        Epijqij_lin_save(:,tt) = Epijqij(nlm, Ecc_opt_lin, Eca_opt_lin, alpha_opt_lin, 1)
        
        ! Nonlinear (n'=3) Sachs enhancements
        Eeiej_nlin_save(:,:,tt) = Eeiej(  nlm, Ecc_opt_nlin, Eca_opt_nlin, alpha_opt_nlin, 3)
        Epijqij_nlin_save(:,tt) = Epijqij(nlm, Ecc_opt_nlin, Eca_opt_nlin, alpha_opt_nlin, 3)

    end
    
    subroutine check(status)
        implicit none
        integer, intent (in) :: status
        if(status /= nf90_noerr) then 
            print *, trim(nf90_strerror(status))
            stop "Stopped"
        end if
    end subroutine check  
    
end program

